### ISSUE TITLE
Silent overflow in absolute()

### ISSUE THREAD

**Issue by endolith** [numpy/numpy#5657]
Python:

```
In [117]: abs(-2147483646)
Out[117]: 2147483646

In [118]: abs(-2147483647)
Out[118]: 2147483647

In [119]: abs(-2147483648)
Out[119]: 2147483648L

In [120]: abs(-2147483649)
Out[120]: 2147483649L

In [121]: abs(-2147483650)
Out[121]: 2147483650L
```

NumPy:

```
In [93]: absolute(array([-2147483646]))
Out[93]: array([2147483646])

In [94]: absolute(array([-2147483647]))
Out[94]: array([2147483647])

In [95]: absolute(array([-2147483648]))
Out[95]: array([-2147483648])

In [96]: absolute(array([-2147483649]))
Out[96]: array([2147483649], dtype=int64)

In [97]: absolute(array([-2147483650]))
Out[97]: array([2147483650], dtype=int64)
```

[In C this is undefined](http://stackoverflow.com/q/11243014/125507) because abs() returns an int and there's no +2147483648, but NumPy should convert to int64?  and then `absolute(array(-2**63))` should convert to object array with longs instead of returning -9223372036854775808?

Related to https://github.com/numpy/numpy/issues/593?

**Comment by matthew-brett** [comment]
There was some discussion of this behavior back in the day: http://numpy-discussion.10968.n7.nabble.com/abs-for-max-negative-integers-desired-behavior-td8939.html

At the time Chuck floated the idea of returning an unsigned int, but was worried about the fact that combining this result with the original would result in an upcast.

Changing the return result type of abs would also be a surprising API change.

On the other hand, the current behavior makes abs a fragile beast of int data types, for which you run the constant risk of returning a large negative number.

I think returning an unsigned int is the correct solution in the long term, so then the question is how to get there.

Maybe one way would be to add a `uabs` function with the desired behavior.    It would differ from abs only in returning an unsigned int for int inputs (and therefore never returning negative numbers).  I would personally vote for also adding a 'always_positive' flag to the `abs` function which would then delegate to `abs`, and a futurewarning about future changes to the uabs behavior as the default.

**Comment by eric-wieser** [comment]
Can't we make the same claim about:

* `np.negative(x)` sometimes doesn't change the sign of x
* `np.add(x, 1)` sometimes flips the sign of x from negative to positive

It seems like this is just an inherent problem with fixed-width types, and not something that can be worked around.

Is there any point introducing `uabs`? It can already be spelt `abs(x).view(np.uint64)`

**Comment by matthew-brett** [comment]
I think uabs would be very helpful.  I've hit this myself and it was very surprising.  Having something at the top of the `abs` docstring pointing to `uabs` would have been very helpful.

Also - `uabs` could be a little more complicated, to preserve the bit-width of the integer (int32 - > uint32, etc).

**Comment by eric-wieser** [comment]
> Also - uabs could be a little more complicated, to preserve the bit-width of the integer (int32 - > uint32, etc).

I think it would be nice to add `np.signed_type(np.uint32) -> np.int32`, `np.unsigned_type(np.int8) -> np.uint8`, which would mostly solve that problem.

Having said that, the current implementation of `abs` negates MIN_INT, which I suppose is undefined behaviour?

**Comment by matthew-brett** [comment]
Eric - I don't think it can be fixed without a change to the return type, because (e.g) int8 can't store +128:

```
np.abs(np.int8(-128))
```

So the return type has to change from int8 to something else to avoid this horror.

**Comment by eric-wieser** [comment]
I agree, but am just pointing out that the user can already request that by doing a cast to the appropriate uint. The alternative is rampant type-promotion, which leads to floats if you're not careful:

```
def unsigned_type(dtype):
    if np.issubdtype(dtype, np.integer):
        return np.dtype(dtype.char.upper())
    return dtype

def uabs(x):
    x = np.asarray(x)
    return np.absolute(x).view(unsigned_type(x.dtype))

a = np.int8(1)
uabs(uabs(uabs(uabs(a) - 1) - 1) - 1) # float64!
```

**Comment by matthew-brett** [comment]
Sure - but I think that risk (of type promotion when using uints) is way less bad than returning a negative number from `(u)abs`...

**Comment by njsmith** [comment]
It's awful, but all integer overflow cases are awful; is this one so much worse that it needs special cases like this? (I don't know the answer. Having proper integer overflow detection in general would certainly help some in this case and many others, though.)

**Comment by matthew-brett** [comment]
Well - I think this case is surprising enough, that when the poor user finally works out what's going on, they'll be thinking "surely there's some simple way of avoiding that".  And there is - it's the `uabs` function.
